cgen: `-autofree` fails

### Describe the bug

The code can work without `-autofree` but if you try to compile with `-autofree`, your code will not be compiled.

```v
import strconv

fn color_code_to_rgb(color string) []int {
	clr := color.replace('#', '')
	return [int(strconv.parse_int(clr[0..2], 16, 0) or { return [0, 0, 0] }),
		int(strconv.parse_int(clr[2..4], 16, 0) or { return [0, 0, 0] }),
		int(strconv.parse_int(clr[4..6],
			16, 0) or { return [0, 0, 0] })]
}

println(color_code_to_rgb('#abcdef'))
```

### Expected Behavior

Works fine and writes the output to console
```v
[171, 205, 239]
```

### Current Behavior

Compiler fails with
```bash
[malisipi@malisipi-20351 ~]$ v -cg -autofree free.v 
/tmp/v_1000/free.11594424565698986927.tmp.c:12879: error: cannot convert 'typedef struct string' to 'int'
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

Compile the code with `v -autofree file_name.v`

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 0a8a0fd

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 3197ec1.0a8a0fd
OS: linux, "Garuda Linux"
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-4210U CPU @ 1.70GHz

getwd: /home/malisipi
vexe: /home/malisipi/Programlar/v/v
vexe mtime: 2023-03-25 13:43:23

vroot: OK, value: /home/malisipi/Programlar/v
VMODULES: OK, value: /home/malisipi/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.39.2
Git vroot status: weekly.2023.05-256-g0a8a0fdb-dirty
.git/config present: true

CC version: cc (GCC) 12.2.1 20230201
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3121,7 +3121,12 @@ fn (mut g Gen) expr(node_ ast.Expr) {
 					g.writeln('(${shared_styp}*)__dup${shared_styp}(&(${shared_styp}){.mtx = {0}, .val =')
 				}
 			}
-			last_stmt_pos := if g.stmt_path_pos.len > 0 { g.stmt_path_pos.last() } else { 0 }
+			stmt_before_call_expr_pos := if g.stmt_path_pos.len > 0 {
+				g.stmt_path_pos.last()
+			} else {
+				0
+			}
+
 			g.call_expr(node)
 			if g.is_autofree && !g.is_builtin_mod && !g.is_js_call && g.strs_to_free0.len == 0
 				&& !g.inside_lambda {
@@ -3130,7 +3135,8 @@ fn (mut g Gen) expr(node_ ast.Expr) {
 				// so just skip it
 				g.autofree_call_pregen(node)
 				if g.strs_to_free0.len > 0 {
-					g.insert_at(last_stmt_pos, g.strs_to_free0.join('\n') + '/* inserted before */')
+					g.insert_at(stmt_before_call_expr_pos, g.strs_to_free0.join('\n') +
+						'/* inserted before */')
 				}
 				g.strs_to_free0 = []
 			}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2021,7 +2021,7 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 					fn_name := node.name.replace('.', '_')
 					// name := '_tt${g.tmp_count_af}_arg_expr_${fn_name}_$i'
 					name := '_arg_expr_${fn_name}_${i + 1}_${node.pos.pos}'
-					g.write('/*af arg*/' + name)
+					g.write('/*autofree arg*/' + name)
 				}
 			} else {
 				g.ref_or_deref_arg(arg, expected_types[i], node.language)

--- a/vlib/v/gen/c/text_manipulation.v
+++ b/vlib/v/gen/c/text_manipulation.v
@@ -91,7 +91,7 @@ fn (mut g Gen) insert_at(pos int, s string) {
 	// in the correct positions, considering the surgically made changes.
 	for index, stmt_pos in g.stmt_path_pos {
 		if stmt_pos >= pos {
-			g.stmt_path_pos[index] = stmt_pos + s.len + 1
+			g.stmt_path_pos[index] += s.len + 1
 		}
 	}
 }

--- a/vlib/v/gen/c/text_manipulation.v
+++ b/vlib/v/gen/c/text_manipulation.v
@@ -84,4 +84,14 @@ fn (mut g Gen) insert_at(pos int, s string) {
 	// g.out_parallel[g.out_idx].cut_to(pos)
 	g.writeln(s)
 	g.write(cur_line)
+
+	// After modifying the code in the buffer, we need to adjust the positions of the statements
+	// to account for the added line of code.
+	// This is necessary to ensure that autofree can properly insert string declarations
+	// in the correct positions, considering the surgically made changes.
+	for index, stmt_pos in g.stmt_path_pos {
+		if stmt_pos >= pos {
+			g.stmt_path_pos[index] = stmt_pos + s.len + 1
+		}
+	}
 }

--- a/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
+++ b/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
@@ -1,0 +1,13 @@
+import strconv
+
+fn color_code_to_rgb(color string) []int {
+	clr := color.replace('#', '')
+	return [int(strconv.parse_int(clr[0..2], 16, 0) or { return [0, 0, 0] }),
+		int(strconv.parse_int(clr[2..4], 16, 0) or { return [0, 0, 0] }),
+		int(strconv.parse_int(clr[4..6],
+			16, 0) or { return [0, 0, 0] })]
+}
+
+fn main() {
+	dump(color_code_to_rgb('#abcdef'))
+}

--- a/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
+++ b/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
@@ -1,13 +1,17 @@
-import strconv
+fn copy_string(s string) !string {
+	return s
+}
 
-fn color_code_to_rgb(color string) []int {
-	clr := color.replace('#', '')
-	return [int(strconv.parse_int(clr[0..2], 16, 0) or { return [0, 0, 0] }),
-		int(strconv.parse_int(clr[2..4], 16, 0) or { return [0, 0, 0] }),
-		int(strconv.parse_int(clr[4..6],
-			16, 0) or { return [0, 0, 0] })]
+// get_hex_rgb_colors returns hex codes of RGB colors
+// with redundant cloning of strings to check autofree is working.
+fn get_hex_rgb_colors(color string) []string {
+	colors := color.replace('#', '')
+
+	return [copy_string(colors[0..2]) or { return ['', '', ''] },
+		copy_string(colors[2..4]) or { return ['', '', ''] },
+		copy_string(colors[4..6]) or { return ['', '', ''] }]
 }
 
 fn main() {
-	dump(color_code_to_rgb('#abcdef'))
+	dump(get_hex_rgb_colors('#abcdef'))
 }

--- a/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
+++ b/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
@@ -1,4 +1,4 @@
-fn copy_string(s string) !string {
+fn return_string(s string) !string {
 	return s
 }
 
@@ -7,9 +7,9 @@ fn copy_string(s string) !string {
 fn get_hex_rgb_colors(color string) []string {
 	colors := color.replace('#', '')
 
-	return [copy_string(colors[0..2]) or { return ['', '', ''] },
-		copy_string(colors[2..4]) or { return ['', '', ''] },
-		copy_string(colors[4..6]) or { return ['', '', ''] }]
+	return [return_string(colors[0..2]) or { return ['', '', ''] },
+		return_string(colors[2..4]) or { return ['', '', ''] },
+		return_string(colors[4..6]) or { return ['', '', ''] }]
 }
 
 fn main() {

--- a/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
+++ b/vlib/v/slow_tests/valgrind/multiple_fn_calls.v
@@ -1,17 +1,13 @@
-fn return_string(s string) !string {
-	return s
-}
+import strconv
 
-// get_hex_rgb_colors returns hex codes of RGB colors
-// with redundant cloning of strings to check autofree is working.
-fn get_hex_rgb_colors(color string) []string {
-	colors := color.replace('#', '')
-
-	return [return_string(colors[0..2]) or { return ['', '', ''] },
-		return_string(colors[2..4]) or { return ['', '', ''] },
-		return_string(colors[4..6]) or { return ['', '', ''] }]
+fn color_code_to_rgb(color string) []int {
+	clr := color.replace('#', '')
+	return [int(strconv.parse_int(clr[0..2], 16, 0) or { return [0, 0, 0] }),
+		int(strconv.parse_int(clr[2..4], 16, 0) or { return [0, 0, 0] }),
+		int(strconv.parse_int(clr[4..6],
+			16, 0) or { return [0, 0, 0] })]
 }
 
 fn main() {
-	dump(get_hex_rgb_colors('#abcdef'))
+	dump(color_code_to_rgb('#abcdef'))
 }

--- a/vlib/v/slow_tests/valgrind/valgrind_test.v
+++ b/vlib/v/slow_tests/valgrind/valgrind_test.v
@@ -27,6 +27,7 @@ const skip_valgrind_files = [
 	'vlib/v/slow_tests/valgrind/struct_field.v',
 	'vlib/v/slow_tests/valgrind/fn_returning_string_param.v',
 	'vlib/v/slow_tests/valgrind/fn_with_return_should_free_local_vars.v',
+	'vlib/v/slow_tests/valgrind/multiple_fn_calls.v',
 	'vlib/v/slow_tests/valgrind/option_simple.v',
 	'vlib/v/slow_tests/valgrind/string_plus_string_plus.v',
 	'vlib/v/slow_tests/valgrind/import_x_json2.v',

